File: gnsals0904/level1/대충_만든_자판.java
import java.util.*;

class Solution {
    int[] minCntAlpha = new int[26];
    int maxLengthKey = 0;

    public int[] solution(String[] keymap, String[] targets) {
        Arrays.fill(minCntAlpha, 0);

        for (int i = 0; i < keymap.length; i++) {
            maxLengthKey = Math.max(maxLengthKey, keymap[i].length());
        }

        for (int i = 0; i < maxLengthKey; i++) {
            for (int j = 0; j < keymap.length; j++) {
                if (keymap[j].length() <= i) {
                    continue;
                }

                int nowKey = keymap[j].charAt(i) - 'A';
                if (minCntAlpha[nowKey] == 0) {
                    minCntAlpha[nowKey] = i + 1;
                }
            }
        }

        int[] answer = new int[targets.length];

        for (int j = 0; j < targets.length; j++) {
            String target = targets[j];
            int result = 0;

            for (int i = 0; i < target.length(); i++) {
                int minCnt = minCntAlpha[target.charAt(i) - 'A'];
                if (minCnt == 0) {
                    result = -1;
                    break;
                }
                result += minCnt;
            }

            answer[j] = result;
        }
        return answer;
    }
}

File: yujin-level0/Week64/대충 만든 자판.js
function solution(keymap, targets) {
  const minPressMap = new Map();

  // 각 문자를 입력하기 위한 최소 누름 횟수를 Map에 계산하여 저장
  for (const key of keymap) {
    for (let i = 0; i < key.length; i++) {
      const char = key[i];
      const pressCount = i + 1; // 인덱스는 0부터 시작하므로 +1

      if (!minPressMap.has(char)) {
        minPressMap.set(char, pressCount);
      } else {
        minPressMap.set(char, Math.min(minPressMap.get(char), pressCount));
      }
    }
  }

  // targets의 각 문자열을 만들기 위한 최소 횟수 계산
  const answer = targets.map((target) => {
    let totalPresses = 0;

    for (const char of target) {
      // 자판으로 입력할 수 없는 문자가 포함된 경우 -1 반환
      if (!minPressMap.has(char)) {
        return -1;
      }
      totalPresses += minPressMap.get(char);
    }

    return totalPresses;
  });

  return answer;
}
